File: csharp/src/Connections/MultiModeTunnelClient.cs
// <copyright file="MultiModeTunnelClient.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// </copyright>

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.VisualStudio.Ssh.Tcp.Events;
using Microsoft.VsSaaS.TunnelService.Contracts;

namespace Microsoft.VsSaaS.TunnelService
{
    /// <summary>
    /// Tunnel client implementation that selects one of multiple available connection modes.
    /// </summary>
    public class MultiModeTunnelClient : ITunnelClient
    {
        /// <summary>
        /// Creates a new instance of the <see cref="MultiModeTunnelClient" /> class
        /// that can select from among multiple single-mode clients.
        /// </summary>
        public MultiModeTunnelClient(IEnumerable<ITunnelClient> clients)
        {
            Clients = new List<ITunnelClient>(Requires.NotNull(clients, nameof(clients)));
            Requires.Argument(
                Clients.Count() > 0,
                nameof(clients),
                "At least one tunnel client is required.");
        }

        /// <summary>
        /// Gets the list of clients that may be used to connect to the tunnel.
        /// </summary>
        public IEnumerable<ITunnelClient> Clients { get; }

        /// <inheritdoc />
        public IReadOnlyCollection<TunnelConnectionMode> ConnectionModes
            => Clients.SelectMany((c) => c.ConnectionModes).Distinct().ToArray();

        /// <inheritdoc />
        public ForwardedPortsCollection? ForwardedPorts => throw new NotImplementedException();

        /// <inheritdoc />
        public bool AcceptLocalConnectionsForForwardedPorts
        {
            get => Clients.Any(c => c.AcceptLocalConnectionsForForwardedPorts);
            set
            {
                foreach (var client in Clients)
                {
                    client.AcceptLocalConnectionsForForwardedPorts = value;
                }
            }
        }

        /// <inheritdoc />
        public async Task ConnectAsync(
            Tunnel tunnel,
            string? hostId,
            CancellationToken cancellation)
        {
            Requires.NotNull(tunnel, nameof(tunnel));

            // TODO: Filter tunnel endpoints by host ID, if specified.
            // TODO: Match tunnel endpoints to client connection modes.

            await Task.CompletedTask;
            throw new NotImplementedException();
        }

        /// <inheritdoc />
        public Task WaitForForwardedPortAsync(int forwardedPort, CancellationToken cancellation)
        {
            throw new NotImplementedException();
        }

        /// <inheritdoc />
        public async ValueTask DisposeAsync()
        {
            var disposeTasks = new List<Task>();

            foreach (var client in Clients)
            {
                disposeTasks.Add(client.DisposeAsync().AsTask());
            }

            await Task.WhenAll(disposeTasks);
        }

        /// <inheritdoc />
        public Task<Stream?> ConnectToForwardedPortAsync(int forwardedPort, CancellationToken cancellation)
        {
            throw new NotImplementedException();
        }
    }
}

File: ts/src/connections/multiModeTunnelClient.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { TunnelConnectionMode, Tunnel, TunnelAccessScopes } from '@microsoft/dev-tunnels-contracts';
import { CancellationToken, SshStream } from '@microsoft/dev-tunnels-ssh';
import { ForwardedPortsCollection } from '@microsoft/dev-tunnels-ssh-tcp';
import { TunnelClient } from './tunnelClient'
import { PortForwardingEventArgs } from './portForwardingEventArgs';
import { TunnelConnectionBase } from './tunnelConnectionBase';
import { TunnelConnectionOptions } from './tunnelConnectionOptions';
import { Event } from 'vscode-jsonrpc';

/**
 * Tunnel client implementation that selects one of multiple available connection modes.
 */
export class MultiModeTunnelClient extends TunnelConnectionBase implements TunnelClient {
    public forwardedPorts: ForwardedPortsCollection | undefined;
    public clients: TunnelClient[] = [];

    public connectionModes: TunnelConnectionMode[] = this.clients
        ? [...new Set(...this.clients.map((c) => c.connectionModes))]
        : [];

    public constructor() {
        super(TunnelAccessScopes.Connect);
    }

    /**
     * A value indicating whether local connections for forwarded ports are accepted.
     * Local connections are not accepted if the host process is not NodeJS (e.g. browser).
     */
    public get acceptLocalConnectionsForForwardedPorts(): boolean {
        return !!this.clients.find((c) => c.acceptLocalConnectionsForForwardedPorts);
    }

    public set acceptLocalConnectionsForForwardedPorts(value: boolean) {
        this.clients.forEach((c) => (c.acceptLocalConnectionsForForwardedPorts = value));
    }

    public get localForwardingHostAddress(): string {
        return this.clients[0]?.localForwardingHostAddress;
    }

    public set localForwardingHostAddress(value: string) {
        this.clients.forEach((c) => (c.localForwardingHostAddress = value));
    }

    public connect(
        tunnel: Tunnel,
        options?: TunnelConnectionOptions,
        cancellation?: CancellationToken,
    ): Promise<void> {
        if (!tunnel) {
            throw new Error('Tunnel cannot be null');
        }

        return new Promise<void>((resolve) => {});
    }

    public get portForwarding() : Event<PortForwardingEventArgs> {
        throw new Error('Not supported.');
    }

    public connectToForwardedPort(
        fowardedPort: number,
        cancellation?: CancellationToken,
    ): Promise<SshStream> {
        throw new Error('Method not implemented.');
    }
    public waitForForwardedPort(
        forwardedPort: number,
        cancellation?: CancellationToken,
    ): Promise<void> {
        throw new Error('Method not implemented.');
    }

    public async refreshPorts(): Promise<void> {
        throw new Error('Method not implemented.');
    }

    public async dispose(): Promise<void> {
        await super.dispose();
        await Promise.all(this.clients.map((client) => client.dispose()));
    }
}
